### PR TITLE
Update windows VS detection code to account for new directory

### PR DESCRIPTION
**Pull request by redsun82**
Windows VS 2022 version 17.6 introduced a new `vcpkg` directory underneath `VC` that is throwing off the toolchain autodetection code.

The checks now got renamed to a more approriate `_is_vs_2017_or_newer` and takes into account the possible presence of this `vcpkg` directory.

Closes https://github.com/bazelbuild/bazel/issues/18592

### CODE REVIEW

**Comment by redsun82** [comment]
Tested this successfully on a recent VS 2022 Community install

**Comment by Mizux** [comment]
Should you also change https://github.com/bazelbuild/bazel/blob/75334d1210dabaaad75d4048e6c7a37fa7504d57/tools/cpp/windows_cc_configure.bzl#L232-L244 ?

ps: I'm using "Microsoft Visual Studio\\2022\\Preview\\VC" (2022 Community preview 17.7.0)

**Comment by redsun82** [comment]
Good catch @Mizux, but now it is kinda surprising me that even though `2022` is not on that list, it still gets picked up :thinking:

**Comment by redsun82** [comment]
Ah, I see now, that is a kind of last resort hard-coded path check, following many other ways of finding out the VC path, which probably usually already work. But yes, then it makes sense to add 2022 to the mix there.

File: tools/cpp/windows_cc_configure.bzl [hunk 1]
```diff
@@ -232,16 +232,12 @@ def find_vc_path(repository_ctx):
...
-        "Microsoft Visual Studio\\2019\\BuildTools\\VC",
-        "Microsoft Visual Studio\\2019\\Community\\VC",
-        "Microsoft Visual Studio\\2019\\Professional\\VC",
-        "Microsoft Visual Studio\\2019\\Enterprise\\VC",
-        "Microsoft Visual Studio\\2017\\BuildTools\\VC",
-        "Microsoft Visual Studio\\2017\\Community\\VC",
-        "Microsoft Visual Studio\\2017\\Professional\\VC",
-        "Microsoft Visual Studio\\2017\\Enterprise\\VC",
         "Microsoft Visual Studio 14.0\\VC",
+        "Microsoft Visual Studio\\2019\\Preview\\VC",
```
Comment (Mizux) [inline tools/cpp/windows_cc_configure.bzl:236]: Currently we need 2022\Preview\VC to build google/or-tools ^^;
or you could add it to below comprehension list.
Also order may matter i.e. prefer "Preview" before Enterprise before Community etc...

Comment (redsun82) [reply tools/cpp/windows_cc_configure.bzl:236]: in previous code the order was "Preview", "BuildTools", "Community", "Professional", "Enterprise". Should I keep that?

Comment (redsun82) [reply tools/cpp/windows_cc_configure.bzl:236]: Also, I guess adding `"Microsoft Visual Studio\\2017\\Preview\\VC"` is harmless even though that did not exist AFAIK, right?

Comment (redsun82) [reply tools/cpp/windows_cc_configure.bzl]: (for the moment I preserved the order prior to this patch wrt editions, and added the 2017\Preview one, but that can still be tweaked)

File: tools/cpp/windows_cc_configure.bzl [hunk 2]
```diff
@@ -254,17 +249,21 @@ def find_vc_path(repository_ctx):
...
-    # The layout of VC folder in VS 2017 and 2019 is different from that in VS 2015 and older versions.
-    # In VS 2017 and 2019, it contains only three directories:
+    # The layout of VC folder in VS 2017 and newer versions is different from that in VS 2015 and older versions.
+    # From VS 2017 it contains three directories:
     # "Auxiliary", "Redist", "Tools"
+    # From VS 2022 17.6, a fourth "vcpkg" directory is also present
 
-    vc_2017_or_2019_contents = ["auxiliary", "redist", "tools"]
+    vc_2017_or_newer_contents = ["auxiliary", "redist", "tools"]
     vc_path_contents = [d.basename.lower() for d in repository_ctx.path(vc_path).readdir()]
```
Comment (mattip) [inline tools/cpp/windows_cc_configure.bzl:261]: Maybe a different approach would be to future-proof this check:
```suggestion
    vc_path_contents = [d.basename.lower() for d in repository_ctx.path(vc_path).readdir() if d.basename.lower() in vc_2017_or_2019_contents]
```

Comment (redsun82) [reply tools/cpp/windows_cc_configure.bzl:261]: I was thinking about it, although I did not have a 2015 VS installation to rule out that it does not contain those entries plus something else. But that could be a way to future-proof this check, yes.

Comment (redsun82) [reply tools/cpp/windows_cc_configure.bzl:262]: done (I've double checked this will not break down on a 2015 VC)

**Comment by redsun82** [comment]
@Mizux 
> Should you also change
> 
> https://github.com/bazelbuild/bazel/blob/75334d1210dabaaad75d4048e6c7a37fa7504d57/tools/cpp/windows_cc_configure.bzl#L232-L244
> 
> ? (to add all 2022 VS C++ flavors)
> ps: I'm using "Microsoft Visual Studio\2022\Preview\VC" (2022 Community preview 17.7.0)

This change makes `//src/test/py/bazel:bazel_windows_cpp_test` [fail](https://buildkite.com/bazel/bazel-bazel-github-presubmit/builds/16261#0188c3b3-720c-40b4-b5ac-bc5d4f3c508b) for a reason I can't understand (those tests fail on my local machine already on master, so I'm having difficulties debugging that). [This draft PR](https://github.com/bazelbuild/bazel/pull/18705) shows that reverting just that change makes the test happy. I'm not sure I can devote a lot of time debugging/fixing that test.

**Comment by lblk** [comment]
@redsun82 can u publish an release binary file for bazel of your branch ? just in you repo's release and name it bazel-vs2022.exe is ok. i'need it for some special work. thank u very much

**Comment by Vertexwahn** [comment]
Can we work around this issue by setting the `BAZEL_VC` environment variable?

**Comment by mattip** [comment]
> Can we work around this issue by setting the BAZEL_VC environment variable?

I did not find one after long hours of debugging. There is no way around the failure other than:
- remove the offending extra directory
- fix the code to accept the new layout

**Comment by lblk** [comment]
> Can we work around this issue by setting the environment variable?`BAZEL_VC`

this will not work on vs2022

**Comment by redsun82** [comment]
> @redsun82 can u publish an release binary file for bazel of your branch ? just in you repo's release and name it bazel-vs2022.exe is ok. i'need it for some special work. thank u very much

@lblk  I'll launch the release build on my machine and let you know

**Comment by redsun82** [comment]
As already noted, setting `BAZEL_VC` won't work. Another workaround that I just tested consists in
* grabbing the [patched `tools/cpp` directory](https://github.com/redsun82/bazel/tree/fix-vs-autodetection/tools/cpp) (you can grab a zip here: https://download-directory.github.io/?url=https%3A%2F%2Fgithub.com%2Fredsun82%2Fbazel%2Ftree%2Ffix-vs-autodetection%2Ftools%2Fcpp)
* dump it into your project
* go into `tools/cpp/cc_configure.bzl` and replace all `@bazel_tools` with nothing
* add the following to your `WORKSPACE`
```bazel
load("//tools/cpp:cc_configure.bzl", "cc_configure")

cc_configure()
```
This will replace the built-in autodetection with the patched one.

**Comment by redsun82** [comment]
@lblk here you go:

https://github.com/redsun82/bazel/releases/tag/vs-2022-fix-0.1

**Comment by redsun82** [comment]
@Mizux what do you think about the change to the hardcoded VS paths that makes that one test fail? As I mentioned I cannot unfortunately devote much time to that currently, so either someone could take over trying to figure out that test failure, or for the moment we revert that change to quickly get something out that still works most of the time? Before this issue I don't think anyone has yet complained about VS 2022 detection not working, even though it's missing from that hard-coded list.

**Comment by lblk** [comment]
> @lblk here you go:
> 
> https://github.com/redsun82/bazel/releases/tag/vs-2022-fix-0.1

thank you very much! so kind of u!

**Review by meteorcloudy — APPROVED** [review]
Thanks for fixing it, sorry for the late reply!

**Comment by meteorcloudy** [comment]
Let's me check if I can help with the test failure

File: tools/cpp/windows_cc_configure.bzl [hunk 3]
```diff
@@ -254,17 +249,19 @@ def find_vc_path(repository_ctx):
...
     # "Auxiliary", "Redist", "Tools"
+    # From VS 2022 17.6, a fourth "vcpkg" directory is also present, but we only check presence of the three
+    # directories above in case other directories pop up in the future
 
-    vc_2017_or_2019_contents = ["auxiliary", "redist", "tools"]
+    vc_2017_or_newer_contents = ["auxiliary", "redist", "tools"]
     vc_path_contents = [d.basename.lower() for d in repository_ctx.path(vc_path).readdir()]
-    vc_path_contents = sorted(vc_path_contents)
-    return vc_path_contents == vc_2017_or_2019_contents
+    vc_path_contents = sorted([d for d in vc_path_contents if d in vc_2017_or_newer_contents])
```
Comment (meteorcloudy) [inline tools/cpp/windows_cc_configure.bzl:263]: I guess we can simplify the logic here:

```
vc_2017_or_newer_expected_dirs = ["auxiliary", "redist", "tools"]
for dir in vc_2017_or_newer_expected_dirs:
    if not repository_ctx.path(vc_path).get_child(dir).exists:
        return False
return True
```

**Review by meteorcloudy — CHANGES_REQUESTED** [review]

**Comment by meteorcloudy** [comment]
Maybe we can also remove support for VS2015 and prior to simplify the code?

**Comment by meteorcloudy** [comment]
@bazel-io fork 6.3.0

**Comment by mattip** [comment]
> Maybe we can also remove support for VS2015 and prior to simplify the code?

That kind of change is beyond my paygrade, but I think it might be too disruptive on the 5.4 branch. Maybe there are some people using really old VS versions?

**Comment by meteorcloudy** [comment]
Yeah, that's a breaking change, we probably shouldn't do that if we plan to cherry pick the fix back to 6.x branches.

**Comment by SuperKogito** [comment]
When it comes to building Tensorflow from source with VC 2022 here is the result of my debugging:

Although, Bazel does not account for VS2022 yet but the if condition for VS2017 and VS2019 can still do the trick.
The difference between the old version and the new version is the ../Hostx64/.. label. This is lowercase in your case for example and mine too (VS2022). However, [the windows_cc_configure.bz script: line 325](https://github.com/bazelbuild/bazel/blob/9c96529fca4a135c162e35ce2882834b879438fb/tools/cpp/windows_cc_configure.bzl#L325C1-L326C1) looks for an uppercase ../HostX64/.. label.

The solution here is either to edit ../Hostx64/.. label (https://github.com/bazelbuild/bazel/blob/9c96529fca4a135c162e35ce2882834b879438fb/tools/cpp/windows_cc_configure.bzl#L325C1-L326C1) and change the lines to

    if _is_vs_2017_or_2019(repository_ctx, vc_path):
        lib = msvc_vars_x64["%{msvc_env_lib_x64}"]
        full_version = _get_vc_full_version(repository_ctx, vc_path)
        tools_path = "%s\\Tools\\MSVC\\%s\\bin\\Hostx64\\%s" % (vc_path, full_version, target_arch)

        # For native windows(10) on arm64 builds host toolchain runs in an emulated x86 environment
        if not repository_ctx.path(tools_path).exists:
            tools_path = "%s\\Tools\\MSVC\\%s\\bin\\Hostx86\\%s" % (vc_path, full_version, target_arch)
or you rename \bin\\Hostx64\\ to \bin\\HostX64\\

also it is vital that these environment variables point to the right paths:

BAZEL_SH=C:\msys64\usr\bin\bash.exe
BAZEL_VC=C:\Program Files (x86)\Microsoft Visual Studio\2022\BuildTools\VC
BAZEL_VC_FULL_VERSION=14.36.32532
This was tested with:
Windows 10
Bazel 6.2.1
Python 3.9.7
Tensorflow 2.12

Hopefully this helps.

File: tools/cpp/windows_cc_configure.bzl (hunk 3 above)
Comment (Vertexwahn) [reply tools/cpp/windows_cc_configure.bzl:263]: @redsun82 Can you overtake the suggested change - would be nice if we could soon merge this in bring it to Bazel 6.3.0

Comment (Vertexwahn) [reply tools/cpp/windows_cc_configure.bzl:263]: @redsun82 Can you overtake the suggested change - would be nice if we could soon merge this in bring it to Bazel 6.3.0

Comment (Vertexwahn) [reply tools/cpp/windows_cc_configure.bzl:263]: I can fork this and add @meteorcloudy  suggestion if @redsun82 does not have time/resources to work on this further...

Comment (Vertexwahn) [reply tools/cpp/windows_cc_configure.bzl:263]: I create a PR with this change -> https://github.com/bazelbuild/bazel/pull/18945 @meteorcloudy, @redsun82

**Comment by meteorcloudy** [comment]
> you rename \bin\Hostx64\ to \bin\HostX64\

Is this really necessary? File name is supposed to be case insensitive on Windows.

**Comment by SuperKogito** [comment]
> Is this really necessary? File name is supposed to be case insensitive on Windows.
Idk, as I opted for editing the path in the code to avoid messing any dependencies on the path.

**Comment by meteorcloudy** [comment]
https://github.com/bazelbuild/bazel/pull/18945 has been merged and cherry picked back to 6.3.0, so we can close this one.

**Pull request closed without merging**